features/filesystem/bd/IAPBlockDevice.h

```diff
@@ -0,0 +1,235 @@
+/* Copyright (c) 2018, James Wang
+ * SPDX-License-Identifier: Apache-2.0
+
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.
+
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+#ifndef MBED_IAP_BLOCK_DEVICE_H
+#define MBED_IAP_BLOCK_DEVICE_H
+
+#include "BlockDevice.h"
+#include "FlashIAP.h"
+#if defined(DEVICE_FLASH) || defined(DOXYGEN_ONLY)
+/** In Application Programming block device
+ *
+ * Useful for store data in User Flash
+ *
+ * @code
+ * #include "mbed.h"
+ * #include "IAPBlockDevice.h"
+ *
+ * #define USER_FLASH_STORE_START (0x00040000)
+ * #define USER_FLASH_STORE_SIZE  (256 * 1024)
+ *
+ *
+ * int main() {
+ *     FlashIAP flash;
+ *     flash.init();
+ *     const uint32_t flash_end = flash.get_flash_start() + flash.get_flash_size();
+ *     const uint32_t last_sector_size = flash.get_sector_size(flash_end - 1);
+ * 
+ *     // use the last sector for test
+ *     const uint32_t bd_start_addr = flash_end - last_sector_size;
+ *     const uint32_t bd_size = last_sector_size;
+ *     IAPBlockDevice bd(bd_start_addr, bd_size);
+ * 
+ *     bd.init();
+ *     char * block = static_cast<char*>(malloc(bd_size));
+ *     if (block != 0)
+ *     {
+ *         strcpy(block, "Hello World!\n");
+ *         bd.erase(0, BLOCK_SIZE);
+ *         bd.program(block, 0, BLOCK_SIZE);
+ *         memset(block, 0, bd_size);
+ *         bd.read(block, 0, BLOCK_SIZE);
+ *         printf("%s", block);
+ *         free(block);
+ *         block = 0;
+ *     }
+ *     bd.deinit();
+ * }
+ * @endcode
+ */
+
+enum
+{
+    BD_ERROR_INVALID_ADDR = -5101,
+    BD_ERROR_INVALID_SIZE = -5102,
+    BD_ERROR_IAP_ERROR = -5103,
+};
+
+class IAPBlockDevice : public BlockDevice
+{
+  public:
+    /** Lifetime of the In Application Programming
+     *
+     * @param addr     Block start addr in User Flash, must be a multiple of Flash program and sector sizes
+     * @param size     Block size in bytes. must be a multiple of Flash program and sector sizes
+     */
+    IAPBlockDevice(uint32_t addr, uint32_t size)
+        : _addr(addr), _size(size)
```

Comment:
I would either add or replace this constructor with one that doesn't get the address and size as parameters, but rather query the flash driver for it (using get_flash_start and get_flash_size). Using a partial portion of the flash can be achieved using the slicing block device.